cedar-policy/src/api.rs

```diff
@@ -2068,6 +2069,61 @@ pub(crate) fn fold_partition<T, A, B, E>(
     Ok((lefts, rights))
 }
 
+/// The "type" of a [`Request`], i.e., the [`EntityTypeName`]s of principal
+/// and resource, and the [`EntityUid`] of action
+#[derive(Debug, Clone, PartialEq, Eq, Hash, PartialOrd, Ord)]
+pub struct RequestEnv {
+    pub(crate) principal: EntityTypeName,
+    pub(crate) action: EntityUid,
+    pub(crate) resource: EntityTypeName,
+}
+
+impl RequestEnv {
+    /// Construct a [`RequestEnv`]
+    pub fn new(principal: EntityTypeName, action: EntityUid, resource: EntityTypeName) -> Self {
+        Self {
+            principal,
+            action,
+            resource,
+        }
+    }
+    /// Get the principal type name
+    pub fn principal(&self) -> &EntityTypeName {
+        &self.principal
+    }
+
+    /// Get the action [`EntityUid`]
+    pub fn action(&self) -> &EntityUid {
+        &self.action
+    }
+
+    /// Get the resource type name
+    pub fn resource(&self) -> &EntityTypeName {
+        &self.resource
+    }
+}
+
+#[doc(hidden)]
+impl From<cedar_policy_validator::types::RequestEnv<'_>> for RequestEnv {
+    fn from(value: cedar_policy_validator::types::RequestEnv<'_>) -> Self {
+        match value {
+            cedar_policy_validator::types::RequestEnv::DeclaredAction {
+                principal,
+                action,
+                resource,
+                ..
+            } => Self {
+                principal: principal.clone().into(),
+                resource: resource.clone().into(),
+                action: action.clone().into(),
+            },
+            //PANIC SAFETY: partial validation should not be enabled and hence `RequestEnv::UndeclaredAction` should not show up
```

Comment:
Since we only use this `From` impl once, I think I'd prefer in-lining it in `get_valid_request_envs` so that this invariant is immediately obvious. As it is, this function looks like it could panic, and I need to know that it's only called in that one place before I'm convinced it's safe.

Reply:
Done.